File: 138-copy-list-with-random-pointer/copy-list-with-random-pointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        if head==None:
            return head
        d={}
        head1=Node(head.val)
        temp1=head1
        temp=head.next
        d[head]=head1
        while temp!=None:
            temp1.next=Node(temp.val)
            temp1=temp1.next
            d[temp]=temp1
            temp=temp.next
        p1=head
        p2=head1
        while p1!=None:
            p2.random=d.get(p1.random,None)
            p2=p2.next
            p1=p1.next
        return head1
        

File: 138-copy-list-with-random-pointer/copy-list-with-random-pointer.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    public Node copyRandomList(Node head) {
        if(head==null){
            return head;
        }
        HashMap<Node,Node> map=new HashMap<>();
        Node temp1=head;
        Node dummy=new Node(-1);
        Node temp2=dummy;
        while(temp1!=null){
            Node newNode=new Node(temp1.val);
            map.put(temp1,newNode);
            temp2.next=newNode;
            temp2=temp2.next;
            temp1=temp1.next;
        }
        Node temp3=head;
        Node temp4=dummy.next;
        while(temp3!=null){
            temp4.random=map.getOrDefault(temp3.random,null);
            temp3=temp3.next;
            temp4=temp4.next;
        }
        return dummy.next;
    }
}